pkg/jobs/scheduled_job_test.go

```diff
@@ -121,3 +125,50 @@ func TestPauseUnpauseJob(t *testing.T) {
 	require.False(t, loaded.IsPaused())
 	require.False(t, loaded.NextRun().Equal(time.Time{}))
 }
+
+// TestScheduleMustHaveClusterVersionAndID tests that the ClusterID and version
+// in the cluster details can be accessed. Further it tests that a schedule
+// cannot be created without these fields. In a mixed version state, however,
+// these fields are always empty.
+func TestScheduleMustHaveClusterVersionAndID(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	ctx := context.Background()
+
+	h, cleanup := newTestHelper(t)
+	defer cleanup()
+
+	sj := h.newScheduledJob(t, "test_job", "test sql")
+	schedules := ScheduledJobDB(h.cfg.DB)
+
+	// Fail without a Cluster version
+	sj.SetScheduleDetails(jobspb.ScheduleDetails{ClusterID: jobstest.DummyClusterID})
+	require.ErrorContains(t, schedules.Create(ctx, sj), "scheduled job created without a cluster version")
+
+	// Fail without a Cluster ID
+	sj.SetScheduleDetails(jobspb.ScheduleDetails{
+		CreationClusterVersion: jobstest.DummyClusterVersion,
+		ClusterID:              uuid.UUID{},
+	})
+	require.ErrorContains(t, schedules.Create(ctx, sj), "scheduled job created without a cluster ID")
+
+	// Succeed with both.
+	sj.SetScheduleDetails(jobstest.AddDummyScheduleDetails(jobspb.ScheduleDetails{}))
+	require.NoError(t, schedules.Create(ctx, sj))
+	loaded := h.loadSchedule(t, sj.ScheduleID())
+	require.EqualValues(t, jobstest.DummyClusterID, loaded.ScheduleDetails().ClusterID)
+	require.EqualValues(t, jobstest.DummyClusterVersion, loaded.ScheduleDetails().CreationClusterVersion)
+
+	// Init a mixed version schedule
+	sjMixed := h.newScheduledJob(t, "mixed_test_job", "test sql")
+
+	// In a mixed version state, clusterID and clusterVersion fields do not populate.
```

Comment:
Hm so this is not actually exercising the `protoutil.Unmarshal` call that a pre-23.2 binary would make on a scheduled job created on 23.2+, which is why I don't think this is enough. This is simply testing that we zero out the fields which as my other comment indicates I'm not sure is needed. When an older version binary attempts to unmarshal protobuf bytes written by a newer version, the added fields will simply be ignored.

Can we maybe replace this with a mixed-version logictest instead?

Reply:
mixed version test added. Given my above comment on version gating the writing of these fields, I still think this test case is necessary.